watcher/diffWatcher.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package watcher
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/sergi/go-diff/diffmatchpatch"
+	"golang.org/x/net/context/ctxhttp"
+)
+
+// DiffWatcher regularly check data at path provided, notifies if changes detected.
+type DiffWatcher struct {
+	latest []byte
+
+	synced []byte
+	diffs  []diffmatchpatch.Diff
+	mu     sync.Mutex
+
+	// Exactly one of url/filepath fields is specified.
+	url      string
+	filepath string
+
+	checkInterval time.Duration
+
+	events chan<- DiffEvent
+}
+
+// DiffEvent refelects diff/error detection.
+type DiffEvent struct {
+	Diffs []diffmatchpatch.Diff
+	Err   error
+}
+
+// NewDiffWatcher is factory for DiffWatcher.
+func NewDiffWatcher(path string, isPathURL bool, checkInterval time.Duration, events chan<- DiffEvent) *DiffWatcher {
+	var w DiffWatcher
+	if isPathURL {
+		w.url = path
+	} else {
+		w.filepath = path
+	}
+	w.checkInterval = checkInterval
+	w.events = events
+
+	w.init()
+	return &w
+}
+
+func (w *DiffWatcher) init() {
```

Comment:
Is this ever going to be called elsewhere? Only one call site, and it seems like a slightly dangerous method...